### ISSUE TITLE
Should read offset be expressed in the same terms as write offset?

### ISSUE THREAD

**Issue by huntc** [rust-embedded-community/embedded-storage#19]
Shouldn't the read offset be expressed in the same terms as the write offset and be a multiple of READ_SIZE?

See https://github.com/rust-embedded-community/embedded-storage/blob/master/src/nor_flash.rs#L14

**Comment by huntc** [comment]
FYI I've implemented https://github.com/nrf-rs/nrf-hal/pull/337 on the assumption that the read offset is a multiple of READ_SIZE. It'd be great to get confirmation on this. Thanks.

**Comment by MathiasKoch** [comment]
@Dirbaio i think you may be qualified to answer this? 🤷

**Comment by Dirbaio** [comment]
Yeah, the current docs are missing specifying that `offset` and `bytes.len()` are multiples of `READ_SIZE`.

However it's true that the majority of chips (all?) have READ_SIZE=1. Maybe we can remove READ_SIZE?

WRITE_SIZE matters especially if the NOR flash is not multiwrite, you don't want impls to do "tricks" behind your back to process unaligned writes because they may write to addresses you don't expect... However for READ_SIZE, if a chip really has a read alignment requirement, impls can read "slightly more" and then give you the subset of bytes, which is fine because reads aren't suppoed to have side effects...

**Comment by huntc** [comment]
> Yeah, the current docs are missing specifying that `offset` and `bytes.len()` are multiples of `READ_SIZE`.
> 
> However it's true that the majority of chips (all?) have READ_SIZE=1. Maybe we can remove READ_SIZE?
> 
> WRITE_SIZE matters especially if the NOR flash is not multiwrite, you don't want impls to do "tricks" behind your back to process unaligned writes because they may write to addresses you don't expect... However for READ_SIZE, if a chip really has a read alignment requirement, impls can read "slightly more" and then give you the subset of bytes, which is fine because reads aren't suppoed to have side effects...

I’m happy to drop read size. However, I’m wondering if there’s still a case for dropping write size for the write buf len… for convenience. 

I’m using Posrcard to serialise what I wish to write. While I pass it in a buffer to serialise, it returns me a slice of that of buffer representing what was serialised. That slice len is byte aligned of course, not Nvmc write aligned in terms of len.

So could we relax the write buf len being write aligned and keep the offset as being write aligned?

**Comment by Dirbaio** [comment]
>  However, I’m wondering if there’s still a case for dropping write size for the write buf len… for convenience.

The idea for the NorFlash is to be a low-level trait where impls just expose the hardware capabilities without extra magic. You need that if you're writing low level stuff such as filesystems, databases, key-value-stores.

If you just want "please write these bytes" you want the `Storage` trait. #14 has an adapter that converts NorFlash to Storage by doing the necessary magic padding and read-modify-writes.

**Comment by huntc** [comment]
Thanks for the explanations. Do you want a PR for the elimination of the READ_SIZE requirement?

**Comment by Dirbaio** [comment]
I'm not sure what's best! :sweat_smile: 

On one hand, chips with READ_SIZE>1 are rare, and in that case the impl could easily do "magic" to simulate READ_SIZE=1.

On the other hand, NorFlash aims to be no-magic lowlevel, and the magic can be done in the `Storage` adapter instead, like for writing, which feels a bit more consistent.
